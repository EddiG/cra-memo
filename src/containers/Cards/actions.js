// @flow
import * as constants from './constants';
import type { Card } from './types';

export const updateCards = (cards: Array<Card>) => ({
  type: constants.UPDATE_CARDS,
  payload: {
    cards,
  },
});

export const closeCards = () => ({
  type: constants.CLOSE_CARDS,
  payload: {},
});

export const triggerCard = (cardId: String) => ({
  type: constants.TRIGGER_CARD,
  payload: {
    id: cardId,
  },
});

export const triggerCardAsync = (cardId: String) => ({
  type: constants.TRIGGER_CARD_ASYNC,
  payload: {
    id: cardId,
  },
});

export const validateCards = () => ({
  type: constants.VALIDATE_CARDS,
  payload: {},
});
