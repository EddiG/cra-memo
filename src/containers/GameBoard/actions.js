import * as constants from './constants';

export const generateCards = () => ({
  type: constants.GENERATE_CARDS,
  payload: {},
});

export const refreshCards = () => ({
  type: constants.REFRESH_CARDS,
  payload: {},
});
