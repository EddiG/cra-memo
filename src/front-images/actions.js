import * as constants from './constants';

export const preloadImagesRequire = () => ({
  type: constants.PRELOAD_FRONT_IMAGES_REQUIRE,
  payload: {},
});

export const preloadImagesComplete = () => ({
  type: constants.PRELOAD_FRONT_IMAGES_COMPLETE,
  payload: {},
});
