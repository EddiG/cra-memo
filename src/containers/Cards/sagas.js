import { delay } from 'redux-saga';
import { take, race, select, call, put, takeEvery } from 'redux-saga/effects';
import { TRIGGER_CARD_ASYNC, VALIDATE_CARDS } from './constants';
import { triggerCard } from './actions';
import { selectOpenedNotGuessedCards } from './selectors';

export function* validateCardsAsync() {
  yield call(delay, 1000);
  yield put({ type: VALIDATE_CARDS });
}

export function* triggerCardAsync(action) {
  const openedCards = yield select(selectOpenedNotGuessedCards);
  if (openedCards.length === 0) {
    yield put(triggerCard(action.payload.id));
  }
  if (openedCards.length === 1) {
    yield put(triggerCard(action.payload.id));
    yield race({
      task: call(validateCardsAsync),
      cancel: take(TRIGGER_CARD_ASYNC),
    });
  }
  if (openedCards.length > 1) {
    yield put({ type: VALIDATE_CARDS });
    yield put(triggerCard(action.payload.id));
  }
}

export default function* watchTriggerCardAsync() {
  yield takeEvery(TRIGGER_CARD_ASYNC, triggerCardAsync);
}
