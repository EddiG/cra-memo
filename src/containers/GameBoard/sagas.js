import { delay } from 'redux-saga';
import { all, select, put, call, takeLatest } from 'redux-saga/effects';
import { updateCards, closeCards } from '../Cards/actions';
import { generateCards as generateNewCards } from './generators';
import { GENERATE_CARDS, REFRESH_CARDS } from './constants';

export function* generateCards() {
  const cards = yield select(generateNewCards);
  yield put(updateCards(cards));
}

export function* refreshCards() {
  yield put(closeCards());
  yield call(delay, 1000);
  yield call(generateCards);
}

export function* watchGenerateCards() {
  yield takeLatest(GENERATE_CARDS, generateCards);
}

export function* watchRefreshCards() {
  yield takeLatest(REFRESH_CARDS, refreshCards);
}

export default function* rootSaga() {
  yield all([watchGenerateCards(), watchRefreshCards()]);
}
