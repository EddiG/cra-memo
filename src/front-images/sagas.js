import { takeLatest, select, call, put } from 'redux-saga/effects';
import { selectImagesSources } from './selectors';
import preload from '../utils/preloader';
import { PRELOAD_FRONT_IMAGES_REQUIRE } from './constants';
import * as actions from './actions';

export function* preloadImages() {
  const images = yield select(selectImagesSources);
  yield call(preload, images);
  yield put(actions.preloadImagesComplete());
}

export default function* watchPreloadImages() {
  yield takeLatest(PRELOAD_FRONT_IMAGES_REQUIRE, preloadImages);
}
