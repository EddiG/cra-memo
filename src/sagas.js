import { fork, put } from 'redux-saga/effects';
import gameBoardSagas from './containers/GameBoard/sagas';
import cardsSagas from './containers/Cards/sagas';
import preloadFrontImages from './front-images/sagas';
import { preloadImagesRequire } from './front-images/actions';
import { generateCards } from './containers/GameBoard/actions';

export default function* rootSaga() {
  yield fork(gameBoardSagas);
  yield fork(cardsSagas);
  yield fork(preloadFrontImages);
  yield put(generateCards());
  yield put(preloadImagesRequire());
}
