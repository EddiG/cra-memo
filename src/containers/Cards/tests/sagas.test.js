import { triggerCardAsync } from '../sagas';
import * as actions from '../actions';

describe('cards saga', () => {
  describe('triggerCardAsync when we have not opened cards', () => {
    const action = actions.triggerCardAsync('card1');
    const generator = triggerCardAsync(action);
    const openedCards = [];

    it('should select opened cards', () => {
      const selectOpenedCardsDesc = generator.next().value;
      expect(selectOpenedCardsDesc).toMatchSnapshot();
    });

    it('should dispatch TRIGGER_CARD', () => {
      const putTriggerCardDesc = generator.next(openedCards).value;
      expect(putTriggerCardDesc).toMatchSnapshot();
    });
  });
  describe('triggerCardAsync when we have one opened card', () => {
    const action = actions.triggerCardAsync('card1');
    const generator = triggerCardAsync(action);
    const openedCards = [{ id: 'card1' }];

    it('should select opened cards', () => {
      const selectOpenedCardsDesc = generator.next().value;
      expect(selectOpenedCardsDesc).toMatchSnapshot();
    });

    it('should dispatch TRIGGER_CARD', () => {
      const putTriggerCardDesc = generator.next(openedCards).value;
      expect(putTriggerCardDesc).toMatchSnapshot();
    });

    it('should start in race validateCardsAsync saga and TRIGGER_CARD_ASYNC', () => {
      const raceValidateCardsDesc = generator.next(openedCards).value;
      expect(raceValidateCardsDesc).toMatchSnapshot();
    });
  });
  describe('triggerCardAsync when we have more than one opened cards', () => {
    const action = actions.triggerCardAsync('card1');
    const generator = triggerCardAsync(action);
    const openedCards = [{ id: 'card1' }, { id: 'card2' }];

    it('should select opened cards', () => {
      const selectOpenedCardsDesc = generator.next().value;
      expect(selectOpenedCardsDesc).toMatchSnapshot();
    });

    it('should dispatch VALIDATE_CARDS', () => {
      const putValidateCardsDesc = generator.next(openedCards).value;
      expect(putValidateCardsDesc).toMatchSnapshot();
    });

    it('should dispatch TRIGGER_CARD', () => {
      const putTriggerCardDesc = generator.next().value;
      expect(putTriggerCardDesc).toMatchSnapshot();
    });
  });
});
