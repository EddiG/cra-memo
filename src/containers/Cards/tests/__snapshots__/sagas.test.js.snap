// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cards saga triggerCardAsync when we have more than one opened cards should dispatch TRIGGER_CARD 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "id": "card1",
      },
      "type": "TRIGGER_CARD",
    },
    "channel": null,
  },
}
`;

exports[`cards saga triggerCardAsync when we have more than one opened cards should dispatch VALIDATE_CARDS 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "VALIDATE_CARDS",
    },
    "channel": null,
  },
}
`;

exports[`cards saga triggerCardAsync when we have more than one opened cards should select opened cards 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`cards saga triggerCardAsync when we have not opened cards should dispatch TRIGGER_CARD 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "id": "card1",
      },
      "type": "TRIGGER_CARD",
    },
    "channel": null,
  },
}
`;

exports[`cards saga triggerCardAsync when we have not opened cards should select opened cards 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`cards saga triggerCardAsync when we have one opened card should dispatch TRIGGER_CARD 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "id": "card1",
      },
      "type": "TRIGGER_CARD",
    },
    "channel": null,
  },
}
`;

exports[`cards saga triggerCardAsync when we have one opened card should select opened cards 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`cards saga triggerCardAsync when we have one opened card should start in race validateCardsAsync saga and TRIGGER_CARD_ASYNC 1`] = `
Object {
  "@@redux-saga/IO": true,
  "RACE": Object {
    "cancel": Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "TRIGGER_CARD_ASYNC",
      },
    },
    "task": Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [],
        "context": null,
        "fn": [Function],
      },
    },
  },
}
`;
